core/arch/arm/kernel/spmc_sp_handler.c

```diff
@@ -0,0 +1,220 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021, Arm Limited
+ */
+#include <assert.h>
+#include <bench.h>
+#include <kernel/panic.h>
+#include <kernel/secure_partition.h>
+#include <kernel/spinlock.h>
+#include <kernel/spmc_sp_handler.h>
+#include <optee_ffa.h>
+#include "thread_private.h"
+
+void spmc_sp_start_thread(struct thread_smc_args *args)
+{
+	thread_sp_alloc_and_run(args);
+}
+
+static void ffa_set_error(struct thread_smc_args *args, uint32_t error)
+{
+	args->a0 = FFA_ERROR;
+	args->a2 = error;
+}
+
+static TEE_Result ffa_get_dst(struct thread_smc_args *args,
+			      struct sp_session *caller,
+			      struct sp_session **dst)
+{
+	if (args->a2 != FFA_PARAM_MBZ) {
+		/* a2 should always be 0 */
+		*dst = NULL;
+		return FFA_INVALID_PARAMETERS;
+	}
+
+	*dst = sp_get_session(FFA_DST(args->a1));
```

Comment:
would be better to load `*dst` only upon returning `FFA_OK`.

Reply:
Done